src/vocabulary/mod.rs

```diff
@@ -0,0 +1,368 @@
+        self.eos_token_id
+    }
+
+    /// Filters out `Prepend` kind of tokenizer's normalizers.
+    fn filter_prepend_normalizers(tokenizer: &mut Tokenizer) {
+        // Main concern is prepend normalizers, for example https://github.com/google/sentencepiece
+        // In `sentencepiece` tokenizer, `▁` is used to denote spaces in the source text,
+        // e.g. `Hello World.` could be tokenized as: [Hello] [▁Wor] [ld] [.]
+        //
+        // We don't want to deal with the special characters, so we remove `Prepend` normalizers.
+        if let Some(normalizer) = tokenizer.get_normalizer() {
+            match normalizer {
+                NormalizerWrapper::Sequence(normalization_sequence) => {
+                    let new_sequence = Sequence::new(
+                        normalization_sequence
+                            .get_normalizers()
+                            .iter()
+                            .filter_map(|normalizer| match normalizer {
+                                NormalizerWrapper::Prepend(_) => None,
+                                _ => Some(normalizer.clone()),
+                            })
+                            .collect(),
+                    );
+                    tokenizer.with_normalizer(new_sequence.into());
+                }
+                NormalizerWrapper::Prepend(_) => {
+                    tokenizer.with_normalizer(None::<NormalizerWrapper>);
+                }
+                _ => {}
+            }
+        }
+    }
+}
+
+impl TryFrom<(&mut Tokenizer, u32)> for Vocabulary {
+    type Error = VocabularyError;
+
+    fn try_from(value: (&mut Tokenizer, u32)) -> Result<Vocabulary, VocabularyError> {
+        let (tokenizer, eos_token_id) = value;
+
+        let mut vocabulary = Vocabulary::new(Some(eos_token_id));
+        for (id, added_token) in tokenizer.get_added_tokens_decoder().iter() {
+            if !added_token.special {
+                vocabulary = vocabulary.insert(added_token.content.clone(), *id);
+            }
+        }
+
+        let processor = TokenProcessor::new(tokenizer)?;
+        for (token, token_id) in tokenizer.get_vocab(false) {
+            let token_bytes = processor.process(token)?;
+            // TODO: lossy is temp:
+            // - in python in was handled by byte_symbol function
+            // - interface needs to be redefined to treat Token type as bytes: Vec<u8>
+            let processed_token = String::from_utf8_lossy(&token_bytes);
+            vocabulary = vocabulary.insert(processed_token, token_id);
+        }
+
+        Ok(vocabulary)
+    }
+}
+
+impl Vocabulary {
+    /// Inserts a token to the vocabulary with the specified identifier.
+    pub fn insert(mut self, token: impl Into<Token>, id: TokenId) -> Vocabulary {
+        self.insert_in_place(token, id);
+        self
+    }
+
+    /// Extends the vocabulary with tokens and their identifiers.
+    pub fn extend<T: Into<Token>, I: IntoIterator<Item = TokenId>>(
```

Comment:
I usually see this kind of method chaining pattern for types with the suffix `Builder`.
It's not necessarily wrong to use it without such a wrapper, but I don't think it adheres to the principle of least surprise.